packages/apollo-engine-reporting/src/plugin.ts

```diff
@@ -0,0 +1,403 @@
+          // support it in the short term
+          const { clientName, clientVersion, clientReferenceId } = clientInfo;
+          // the backend makes the choice of mapping clientName => clientReferenceId if
+          // no custom reference id is provided
+          treeBuilder.trace.clientVersion = clientVersion || '';
+          treeBuilder.trace.clientReferenceId = clientReferenceId || '';
+          treeBuilder.trace.clientName = clientName || '';
+        }
+      }
+
+      let endDone: boolean = false;
+      function didEnd() {
+        if (endDone) return;
+        endDone = true;
+        treeBuilder.stopTiming();
+
+        treeBuilder.trace.fullQueryCacheHit = !!metrics.responseCacheHit;
+        treeBuilder.trace.forbiddenOperation = !!metrics.forbiddenOperation;
+        treeBuilder.trace.registeredOperation = !!metrics.registeredOperation;
+
+        // If the user did not explicitly specify an operation name (which we
+        // would have saved in `executionDidStart`), but the request pipeline made
+        // it far enough to figure out what the operation name must be and store
+        // it on requestContext.operationName, use that name.  (Note that this
+        // depends on the assumption that the RequestContext passed to
+        // requestDidStart, which does not yet have operationName, will be mutated
+        // to add operationName later.)
+        const operationName = requestContext.operationName || '';
+
+        // If this was a federated operation and we're the gateway, add the query plan
+        // to the trace.
+        if (metrics.queryPlanTrace) {
+          treeBuilder.trace.queryPlan = metrics.queryPlanTrace;
+        }
+
+        addTrace({
+          operationName,
+          queryHash: requestContext.queryHash!,
+          documentAST: requestContext.document,
+          queryString,
+          trace: treeBuilder.trace,
+          schemaHash: requestContext.schemaHash,
+        });
+      }
+
+      return {
+        parsingDidStart() {
+          ensurePreflight();
+        },
+
+        validationDidStart() {
+          ensurePreflight();
+        },
+
+        didResolveOperation() {
+          ensurePreflight();
+        },
+
+        executionDidStart() {
+          ensurePreflight();
+          return didEnd;
+        },
+
+        willResolveField(...args) {
+          const [, , , info] = args;
+          return treeBuilder.willResolveField(info);
+          // We could save the error into the trace during the end handler, but
+          // it won't have all the information that graphql-js adds to it later,
+          // like 'locations'.
+        },
+
+        didEncounterErrors({ errors }) {
+          // We don't report some special-cased errors to Graph Manager.
+          // See the definition of this function for the reasons.
+          if (allUnreportableSpecialCasedErrors(errors)) {
+            return;
+          }
```

Comment:
This a backwards compatibility bit to account for the fact that we don't ship APQ miss errors to Graph Manager, but that happens _before_ `parsingDidStart` and after `requestDidStart`.  Furthermore though, unlike the `graphql-extensions` API, the `didEncounterErrors` hook in the plugin API _is_ triggered when there is an APQ miss (or APQ misconfiguration).

Basically, we want this to behave the same as before.